### PR TITLE
Add NeoPixel Illumination plugin

### PR DESCRIPTION
**Pull request by brettvitaz**
Use NeoPixel Illumination to control NeoPixels from the OctoPrint interface using the Raspberry Pi's GPIO. Parses GCODE M150 commands for automated control while printing.

### CODE REVIEW

**Review by cp2004 — CHANGES_REQUESTED** [review]
I don't think that the 'sudo' password should be stored by the plugin, and definitely not exposed in the settings API either. It's a basic security issue, as it allows anyone with access to the OctoPrint UI access to *also* login to the server.

There's a few assumptions used that while they may work well for your installation, it might explode on other configurations - I've attempted to point out as many of them used as I have spotted, but do go through and have a look - hardcoded paths will almost never work.

* Ideally don't store password https://github.com/brettvitaz/OctoPrint-Neopixel_Illumination/blob/b4b9950c8f439870d5ab4dc54be4ebfabefe33c1/octoprint_neopixel_illumination/__init__.py#L89, if really not possible then it should not be exposed on the settings API, use [`get_settings_restricted_paths`](https://docs.octoprint.org/en/master/plugins/mixins.html#octoprint.plugin.SettingsPlugin.get_settings_restricted_paths)
* If `on_settings_save` has been called on your plugin, then one of it's settings has changed - OctoPrint will only call this for your plugin if your plugin's settings have changed. No need to do the hard work twice. https://github.com/brettvitaz/OctoPrint-Neopixel_Illumination/blob/b4b9950c8f439870d5ab4dc54be4ebfabefe33c1/octoprint_neopixel_illumination/__init__.py#L156
* Users can change the username, and may also not have OctoPrint installed in this exact location https://github.com/brettvitaz/OctoPrint-Neopixel_Illumination/blob/b4b9950c8f439870d5ab4dc54be4ebfabefe33c1/octoprint_neopixel_illumination/__init__.py#L204 - use `sys.executable` instead.
* It would be good if `time.sleep` was not called here - long running stuff in `on_after_startup` will significantly slow down the startup time of the server. https://github.com/brettvitaz/OctoPrint-Neopixel_Illumination/blob/b4b9950c8f439870d5ab4dc54be4ebfabefe33c1/octoprint_neopixel_illumination/__init__.py#L208
* The assumption of the log path here will fall apart if people have OctoPrint installed anywhere that has a different username or specifies a different basedir. They may even have specified a different basedir. https://github.com/brettvitaz/OctoPrint-Neopixel_Illumination/blob/b4158d8bc206aacf67dd12c18a55750dc29fe208/octoprint_neopixel_illumination/sock_api.py#L9-L12 - I pass the path through from OctoPrint so it is always set correctly like this using [`self._settings.get_plugin_logfile_path`](https://github.com/cp2004/OctoPrint-WS281x_LED_Status/blob/c98e53cb7f60df8fdae4463cdcdf6aaa38ba70a4/octoprint_ws281x_led_status/__init__.py#L323). That returns the path like `<basedir>/logs/plugin_<plugin id>.log`. If the path doesn't exist, log files should not end with a path of `./` - that could be anywhere on the system, depending on how OctoPrint is being run. 

---

It may seem like we're picky, but it is mostly correcting the assumptions so that people don't come complaining when things go wrong!

It is possible to run WS281x type LEDs without root permissions, using the SPI pins. That's how my plugin [WS281x LED Status](https://github.com/cp2004/OctoPrint-WS281x_LED_Status) does things. I've spent a *long* time working with these kinds of LEDs and OctoPrint, so if you have any questions do let me know and we can get stuff sorted.

**Comment by cp2004** [comment]
And finally, a preview of what it will look like https://cp2004-plugins-octoprint-org-git-pr-1073-cp2004.vercel.app/plugins/neopixel_illumination/

**Comment by brettvitaz** [comment]
Thank you for that thorough code review. I will start looking into these items, and will review the documentation you have created for your own plugin. 

Regarding the SPI pins, one of my Raspberry Pi's SPI pins are non-functional and I did not want to limit users when other options were open. I'm not sure what do to about that, and it is one of the reasons for all of the extra work to create the remote service for controlling the neopixels. Maybe I'll just have to come back to that when a better solution presents itself.

File: _plugins/neopixel_illumination.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+title: OctoPrint-NeoPixel Illumination
+description: Control NeoPixels through color picker and GCODE.
+authors:
+- brettvitaz
+license: AGPLv3
+date: 2022-05-24
+
+homepage: https://github.com/brettvitaz/OctoPrint-Neopixel_Illumination
+source: https://github.com/brettvitaz/OctoPrint-Neopixel_Illumination
+archive: https://github.com/brettvitaz/OctoPrint-Neopixel_Illumination/archive/master.zip
```
[the remark this answers is not in the document]
Comment (brettvitaz) [reply _plugins/neopixel_illumination.md]: Good catch

File: _plugins/neopixel_illumination.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+This plugin will allow control of NeoPixels from OctoPrint.
+
+- Use your Raspberry Pi to power and control your NeoPixels. (I have tested powering up to 25 pixels from a Raspberry Pi 4).
+- Change color and intensity from a color picker dialog.
+- Intercept GCODE [M150][M150] commands and execute them on the Raspberry Pi.
+
+## Configuration
+
+| Setting | Description | Default |
+| :--- | :--- | :--- |
```
[the remark this answers is not in the document]
Comment (brettvitaz) [reply _plugins/neopixel_illumination.md]: Thank you for this.

**Comment by kantlivelong** [comment]
Stale - Closing

**Pull request closed without merging**